Allow subprojects to be built for both the build and host machine

dcbaker opened the pull request:
Currently, subprojects can only be built for the host machine. This is obviously not a great situation, as you may need to build dependencies for the build machine. This series fixes that (hopefully for both cmake and native subprojects), but allowing a subproject to be configured twice.

This is accomplished by creating copies of the core datastructures (Build, Environment, and CoreData), with the per-machine attributes set so that both the host and the build point to the overall build configuration. All install targets from build machine subprojects have been dropped, to avoid such unsupported installation. In addition, I've ensured that in a host == build setup only one subproject will be configured, so that in the case of
```meson
d = dependency('foo', native : false, allow_fallback : true)
d_build = dependency('foo', native : true, allow_fallback : true)
```
that `d_build` is just a reference to `d`

This also requires tracking a number of things on a per-machine basis that weren't before. Namely program overrides. In general we want to override for the host machine, but in some cases it makes sense to override for the build machine (such as for build only subprojects), so a native keyword has been added. This may be useful in cases where different output is needed for the host and build machine anyway.

I've made an effort to split out boilerplate changes (and there are several of them), to make this easier to review.

I won't be surprised if there's still some bugs in this, I've run it through some cross compilation on my system, but it's a large complicated set of changes.

dnicolodi commented on mesonbuild/build.py:
```diff
@@ -241,7 +241,8 @@ def __init__(self, environment: environment.Environment):
         self.project_name = 'name of master project'
```
Commit messages typos: `which fix this. This goal of this patch` should be `which fixes this. The goal of this patch`. Also a full stop missing after the first paragraph, if you like.

Nomura-RH commented:
I tried setting up a simple cross compilation project that builds a native executable, and builds a dependency for the executable from a subproject. It didn't work:

_Top-level meson.build:_
```meson
project('myexecutable', 'c')

subproject('mylib')

mylib_dep = dependency('mylib', native: true)

myexecutable = executable('myexecutable', 'main.c', native: true, dependencies: mylib_dep)

test('myexecutable-test', myexecutable)
```

_subprojects/mylib/meson.build:_
```meson
project('mylib', 'c')

mylib = static_library('mylib', 'main.c')

mylib_dep = declare_dependency(link_with: mylib)

meson.override_dependency('mylib', mylib_dep)
```

_meson log:_
```
The Meson build system
Version: 1.2.0.rc3
Source dir: /c/Users/nomura/Desktop/myexecutable
Build dir: /c/Users/nomura/Desktop/myexecutable/build
Build type: cross build
Project name: myexecutable
Project version: undefined
C compiler for the host machine: /c/Users/nomura/Desktop/myexecutable/tools/mwccarm (mwccarm 3.0.136)
C linker for the host machine: /c/Users/nomura/Desktop/myexecutable/tools/mwldarm mwldarm 2.0.87
C compiler for the build machine: gcc (gcc 11.3.0 "gcc (GCC) 11.3.0")
C linker for the build machine: gcc ld.bfd 2.40
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: arm
Host machine cpu: arm
Target machine cpu family: arm
Target machine cpu: arm

Executing subproject mylib 

mylib| Project name: mylib
mylib| Project version: undefined
mylib| C compiler for the host machine: /c/Users/nomura/Desktop/myexecutable/tools/mwccarm (mwccarm 3.0.136)
mylib| C linker for the host machine: /c/Users/nomura/Desktop/myexecutable/tools/mwldarm mwldarm 2.0.87
mylib| C compiler for the build machine: gcc (gcc 11.3.0 "gcc (GCC) 11.3.0")
mylib| C linker for the build machine: gcc ld.bfd 2.40
mylib| Build targets in project: 1
mylib| Subproject mylib finished.

WARNING: Subproject 'mylib' did not override 'mylib' dependency and no variable name specified
Dependency mylib from subproject subprojects/mylib found: NO

meson.build:5:12: ERROR: Dependency 'mylib' is required but not found.
```

Nomura-RH commented on docs/markdown/snippets/subproject_and_override_program_native.md:
```diff
@@ -0,0 +1,13 @@
+## subproject() and find_program() now support the native keyword argument
```
Doesn't `find_program()` already support the `native` keyword?

dcbaker replied:
Yes, oops. That should be meson.override_find_program

dcbaker commented:
@Nomura-RH did I understand correctly that you did get it working?

Nomura-RH commented:
> @Nomura-RH did I understand correctly that you did get it working?

Yep.

Nomura-RH commented:
Ok, this time I'm pretty sure I've found an error : ) Nothing build-breaking though. 

Sub-subprojects are reported in the logs as being built for the host machine, even when they are actually being built for the build machine. Here is a slightly more involved version of my previous example to show what I mean:

_Top-level meson.build_:
```meson
project('myexecutable', 'c')

mylib_dep = dependency('mylib', native: true)

myexecutable = executable('myexecutable', 'main.c', native: true, dependencies: mylib_dep)

test('myexecutable-test', myexecutable)
```

_subprojects/mylib/meson.build_:
```meson
project('mylib', 'c')

mysublib_dep = dependency('mysublib')

mylib = static_library('mylib', 'main.c', dependencies: mysublib_dep)

mylib_dep = declare_dependency(link_with: mylib)

meson.override_dependency('mylib', mylib_dep)
```

_subprojects/mysublib/meson.build_:
```meson
project('mysublib', 'c')

mysublib = static_library('mysublib', 'main.c')

mysublib_dep = declare_dependency(link_with: mysublib)

meson.override_dependency('mysublib', mysublib_dep)
```

_meson log_:
```
The Meson build system
Version: 1.2.99
Source dir: /c/Users/nomura/Desktop/myexecutable
Build dir: /c/Users/nomura/Desktop/myexecutable/build
Build type: cross build
Project name: myexecutable
Project version: undefined
C compiler for the host machine: /c/Users/nomura/Desktop/myexecutable/tools/mwccarm (mwccarm 3.0.136)
C linker for the host machine: /c/Users/nomura/Desktop/myexecutable/tools/mwldarm mwldarm 2.0.87
C compiler for the build machine: gcc (gcc 11.3.0 "gcc (GCC) 11.3.0")
C linker for the build machine: gcc ld.bfd 2.40
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: arm
Host machine cpu: arm
Target machine cpu family: arm
Target machine cpu: arm
Found pkg-config: /usr/bin/pkg-config (1.9.4)
Did not find CMake 'cmake'
Found CMake: NO
Build-time dependency mylib found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency mylib

Executing subproject mylib for machine: build 

mylib| Project name: mylib (for build machine)
mylib| Project version: undefined
mylib| C compiler for the host machine: gcc (gcc 11.3.0 "gcc (GCC) 11.3.0")
mylib| C linker for the host machine: gcc ld.bfd 2.40
mylib| Found pkg-config: /usr/bin/pkg-config (1.9.4)
mylib| Did not find CMake 'cmake'
mylib| Found CMake: NO
mylib| Run-time dependency mysublib found: NO (tried pkgconfig and cmake)
mylib| Looking for a fallback subproject for the dependency mysublib

Executing subproject mysublib for machine: host

mysublib| Project name: mysublib (for build machine)
mysublib| Project version: undefined
mysublib| C compiler for the host machine: gcc (gcc 11.3.0 "gcc (GCC) 11.3.0")
mysublib| C linker for the host machine: gcc ld.bfd 2.40
mysublib| Build targets in project: 1
mysublib| Subproject mysublib finished.

mylib| Dependency mysublib found: YES undefined (overridden)
mylib| Build targets in project: 2
mylib| Subproject mylib finished.

Dependency mylib found: YES undefined (overridden)
Build targets in project: 3

myexecutable undefined

  Subprojects (for host machine)
    mysublib    : YES

  Subprojects (for build machine)
    mylib       : YES

  User defined options
    Cross files : cross.ini
    Native files: native.ini

Found ninja-1.11.1 at /usr/bin/ninja
```

The log errors are:
```
Executing subproject mysublib for machine: host
```

```
Subprojects (for host machine)
    mysublib    : YES
```

dcbaker commented:
@Nomura-RH oh, good catch. Yeah, because the for-build-only subprojects think that host == build they report that they're building for the host machine. Should be fixed in the latest version (with a test!)

Nomura-RH commented:
> @Nomura-RH oh, good catch. Yeah, because the for-build-only subprojects think that host == build they report that they're building for the host machine. Should be fixed in the latest version (with a test!)

The first error is indeed solved, but the second one is still there (the summary at the end still reports the subsubproject in the list of subprojects for the host machine)

dcbaker commented:
Yup, I didn't pass the change down a level... Fixed now in both cases

Nomura-RH commented:
> Yup, I didn't pass the change down a level... Fixed now in both cases

Almost there :) Now all subprojects get listed for both machines:

```  
Subprojects (for host machine)
    mylib       : YES
    mysublib    : YES (from mylib)

  Subprojects (for build machine)
    mylib       : YES
    mysublib    : YES
```

dcbaker commented:
In the "10 native subproject" I see only the "both" subproject in both, and then "buildtool", "hostp", and "test installs" in the build list. but that calling a subproject from a subproject that was called in the main project is probably worth testing... let me see

dcbaker commented:
I've added more tests for correct machine reporting, and discovered that output rpaths weren't being set correctly in the process, so there's another bug down

dcbaker commented:
Sigh, windows

tristan957 commented:
> When a subproject is built for both machines, we need to ensure that the
outputs don't clash. To do this we'll put the outputs of build only
subprojects in a separate root folder (build.subprojects/). We need
special methods for this.

Any thought on `subprojects/host` and `subprojects/build`?

dcbaker commented:
> Any thought on subprojects/host and subprojects/build?

Mainly it's much easier to inject a prefix than inserting something in the middle of the path. I think we'd end up having to do something like:
 ```python
def get_subdir(self) -> str:
    if self.subproject:
        split = os.path.split(self.subdir)
        split.insert(1, 'build' if self.is_build_only else 'host')
        return os.path.join(split)
    return self.subdir
```

Also, doing this means that the layout of the common case remains the same.

jpakkane commented:
I have not read the implementation but if I were to do cross building again from scratch, it would go something along the lines of:

- in build dir have subdirs `native` and `<crossname>`
- when a project is configured it is either native-only or cross only
- if you need it twice (for example, because of `find_program`), it would be configured again
- if you need multiple cross builds at the same time (which apparently some people doing firmware stuff do need), it could be added trivially

dcbaker commented:
At the moment we only really support having two machines, and adding more than one machine is going to be an extremely invasive set of changes. I'm not going to tackle that in this sires, it's simply too much. It would basically require redesigning Meson, or bolting it on with a module of some kind

This configures a given subproject once for each machine (if the machines are different), and puts the output in two different directories, a subproject for the build machine is placed in `build.{subdir}`, while one for the host machine is placed in just `{subdir}`. This means that for the common case nothing changes, and for the not common case it remains the same.

In theory you could have multiple machines from this approach by just replacing `build.` with something else.

bonzini commented:
I see... I am hoping that your recent typed_kwargs work changes the balance... 🤷 It's been two years so maybe it works better now.

There's new work to do to create a second option store if we go down the copy road, so I don't mind spending some time experimenting on the interpreter.

bonzini commented:
For future record I found some more issues, the most notable one is that projects_args and projects_link_args need two layers of PerMachine-ness: one for the subproject and one from the `native` argument of `add_project_arguments` or similar.

bonzini commented:
Closing in favor of #15404
